### ISSUE TITLE
Pattern-based exceptions in .gitignore not working

### ISSUE THREAD

**Issue by swisspol** [libgit2/libgit2#3224]
Example case:

```
*.data
!dont_ignore/*.data
```

Then add a couple files in the workdir: `foo.data` and `dont_ignore/bar.data`.

`git_diff_index_to_workdir()` will correctly ignore `foo.data` but incorrectly also ignore `dont_ignore/bar.data`.

Note that the `.gitignore` file is working as expected when testing with `git status --ignored`.

**Comment by swisspol** [comment]
Doing a bit more testing, this exception works: `!test.data`.

**Comment by carlosmn** [comment]
It looks like the `*.data` rule satisfies the restriction that negation rules must negate an existing pattern.

```C
void test_status_ignore__deeper(void)
{
   int ignored;

    g_repo = cl_git_sandbox_init("empty_standard_repo");

    cl_git_mkfile("empty_standard_repo/.gitignore",
		  "*.data\n"
		  "!dont_ignore/*.data\n");

    cl_git_pass(p_mkdir("empty_standard_repo/dont_ignore", 0777));
    cl_git_mkfile("empty_standard_repo/foo.data", "");
    cl_git_mkfile("empty_standard_repo/bar.data", "");
    cl_git_mkfile("empty_standard_repo/dont_ignore/foo.data", "");
    cl_git_mkfile("empty_standard_repo/dont_ignore/bar.data", "");

    cl_git_pass(git_ignore_path_is_ignored(&ignored, g_repo, "foo.data"));
    cl_assert_equal_i(1, ignored);
    cl_git_pass(git_ignore_path_is_ignored(&ignored, g_repo, "bar.data"));
    cl_assert_equal_i(1, ignored);

    cl_git_pass(git_ignore_path_is_ignored(&ignored, g_repo, "dont_ignore/foo.data"));
    cl_assert_equal_i(0, ignored);
    cl_git_pass(git_ignore_path_is_ignored(&ignored, g_repo, "dont_ignore/bar.data"));
    cl_assert_equal_i(0, ignored);
}
```

**Comment by fogaztuc** [comment]
Just found this issue as well.

**Comment by overboen** [comment]
I believe that there is an issue in the [Atom tree-view](https://github.com/atom/tree-view) project that is caused by this via [Atom's git-utils'](https://github.com/atom/git-utils) project's dependency on this project. It looks like the project is using v0.24.3 of libgit2.

https://github.com/atom/tree-view/issues/902

**Comment by segevfiner** [comment]
Not sure if this is the same, but in the CPython repository they have this:
```
/python
!/Python/
```

The "Python" directory is ignored by libgit2 but *not* by Git.

Tested on Windows.

**Comment by pks-t** [comment]
@segevfiner I guess your issue is different, as it is caused by issues with case-insensitivity. It would be very nice if you could create a new issue describing the problem.

**Comment by vringar** [comment]
This issue should have been closed when #4296 was merged but wasn't

**Comment by ethomson** [comment]
Thanks @vringar !
